File: regex_extracting/extracting/regex_lib/NetworkModel.py
# encoding=utf8
"""
    Author:  'jdwang'
    Date:    'create date: 2017-01-13'; 'last updated date: 2017-01-13'
    Email:   '383287471@qq.com'
    Describe:
"""
from __future__ import print_function
from regex_extracting.extracting.common.regex_base import RegexBase

__version__ = '1.3'


class NetworkModel(RegexBase):
    name = '网络制式'

    def __init__(self, sentence):
        # 要处理的输入句子

        self.sentence = sentence

        # region 1 初始化正则表达式
        # 描述正则表达式
        self.statement_regexs = [
            '网络制式'
        ]
        # 值正则表达式
        self.value_regexs = [
            '单卡双模|单卡多模|双卡单模|双卡双模|双卡双待|双模双待|移动[2-4]G|联通[2-4]G|电信[2-4]G|移动|联通|电信|[2-4]G网络|[2-4]G',
            '随意|随便|都可以|其他|别的'
        ]
        # endregion
        super(NetworkModel, self).__int__()
        self.regex_process()


if __name__ == '__main__':
    price = NetworkModel('价不高')
    for info_meta_data in price.info_meta_data_list:
        print('-' * 80)
        print(str(info_meta_data))

File: regex_extracting/asp_version/Extracting/NetworkModel.cs
/*
 * author : wxp
 * date : 2013/9/11
 * */

using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using System.Text.RegularExpressions;
using System.Web;

namespace Ch2R.Code.Extracting
{
    public class NetworkModel : ParameterBase
    {
        public NetworkModel(Areas.Chat.Models.UserSession userSession, string sentence)
            : base(userSession, sentence)
        {
            name = "网络制式";

            #region 初始化正则表达式
            // 描述正则表达式
            StatementRegexs = new string[]
            {
                @"网络制式"    
            };
            // 值正则表达式
            ValueRegexs = new string[]
            {
                //by WJD 
                //添加4G网络的匹配
                @"单卡双模|单卡多模|双卡单模|双卡双模|双卡双待|双模双待|移动[2-4]G|联通[2-4]G|电信[2-4]G|移动|联通|电信|[2-4]G网络|[2-4]G",
                @"随意|随便|都可以|其他|别的"
            };
            #endregion

            RegexProcess();
        }

        public override void Format()
        {
            for (int i = 0; i < list.Count; ++i)
            {
                if (Regex.IsMatch(list[i].rawData, @"随意|随便|都可以"))
                {
                    list[i].normalData = "";
                }
                else if (Regex.IsMatch(list[i].rawData, @"别的|其他的"))
                {
                    if (userSession.liveTable.Parameter["网络制式"] != null)
                    {
                        list[i].normalData = (String)userSession.liveTable.Parameter["网络制式"];
                        list[i].isNegative = true;
                    }
                    else
                    {
                        /**
                         * 仍有其他价格为null的情况
                         * 别的未测试
                         * */
                    }
                }
                else
                {
                    list[i].normalData = list[i].rawData;
                }
                /* 这里处理一下否定情况
                 * 做法先只是简单的判断原句中信息元数据之前有没有出现否定语
                 * 比如用户说：我不要移动的手机
                 * 那么匹配上：移动
                 * 在“移动”前出现“不”这个否定词，因此认为是否定形式
                 */
                if (Regex.IsMatch(sentence.Substring(0, list[i].leftIndex), @"(不)|(非(?!常))"))
                {
                    list[i].isNegative = true;
                }
            }
        }

        public override InfoBlock ToInfoBlock()
        {
            List<InfoUnit> L = new List<InfoUnit>();
            foreach (InfoMetadata im in list)
            {
                InfoUnit iu = new InfoUnit();
                iu.normalData = im.normalData;
                iu.rawData = im.rawData;
                iu.isNegative = im.isNegative;
                L.Add(iu);
            }
            InfoBlock ib = new InfoBlock(name, "or", L);
            /* 这里关系应该是“or”
             * 网络制式参数的信息是相互独立的
             */
            return ib;
        }
    }
}